pkg/services/anonymous/anonimpl/impl_test.go

```diff
@@ -58,75 +63,129 @@ func TestAnonDeviceKey(t *testing.T) {
 	}
 }
 
-func TestIntegrationAnonDeviceService_tag(t *testing.T) {
+func TestIntegrationDeviceService_tag(t *testing.T) {
+	type tagReq struct {
+		httpReq *http.Request
+		kind    anonymous.DeviceKind
+	}
 	testCases := []struct {
-		name          string
-		req           []*http.Request
-		expectedCount int64
+		name                string
+		req                 []tagReq
+		expectedAnonCount   int64
+		expectedAuthedCount int64
+		expectedDevice      *Device
 	}{
 		{
-			name:          "no requests",
-			req:           []*http.Request{},
-			expectedCount: 0,
+			name:                "no requests",
+			req:                 []tagReq{{httpReq: &http.Request{}, kind: anonymous.AnonDevice}},
+			expectedAnonCount:   0,
+			expectedAuthedCount: 0,
 		},
 		{
 			name: "missing info should not tag",
-			req: []*http.Request{
-				{
-					Header: http.Header{
-						"User-Agent": []string{"test"},
-					},
+			req: []tagReq{{httpReq: &http.Request{
+				Header: http.Header{
+					"User-Agent": []string{"test"},
 				},
 			},
-			expectedCount: 0,
+				kind: anonymous.AnonDevice,
+			}},
+			expectedAnonCount:   0,
+			expectedAuthedCount: 0,
 		},
 		{
 			name: "should tag once",
-			req: []*http.Request{
-				{
-					Header: http.Header{
-						"User-Agent":      []string{"test"},
-						"X-Forwarded-For": []string{"10.30.30.1"},
-					},
+			req: []tagReq{{httpReq: &http.Request{
+				Header: http.Header{
+					"User-Agent":      []string{"test"},
+					"X-Forwarded-For": []string{"10.30.30.1"},
 				},
 			},
-			expectedCount: 1,
+				kind: anonymous.AnonDevice,
+			},
+			},
+			expectedAnonCount:   1,
+			expectedAuthedCount: 0,
+			expectedDevice: &Device{
+				Kind:      anonymous.AnonDevice,
+				IP:        "10.30.30.1",
+				UserAgent: "test"},
 		},
 		{
 			name: "repeat request should not tag",
-			req: []*http.Request{
-				{
-					Header: http.Header{
-						"User-Agent":      []string{"test"},
-						"X-Forwarded-For": []string{"10.30.30.1"},
-					},
+			req: []tagReq{{httpReq: &http.Request{
+				Header: http.Header{
+					"User-Agent":      []string{"test"},
+					"X-Forwarded-For": []string{"10.30.30.1"},
+				},
+			},
+				kind: anonymous.AnonDevice,
+			}, {httpReq: &http.Request{
+				Header: http.Header{
+					"User-Agent":      []string{"test"},
+					"X-Forwarded-For": []string{"10.30.30.1"},
+				},
+			},
+				kind: anonymous.AnonDevice,
+			},
+			},
+			expectedAnonCount:   1,
+			expectedAuthedCount: 0,
+		}, {
+			name: "repeat request should tag if different kind",
```

Comment:
This is an interesting case. I think it's the correct behaviour, but I wonder if we need to explicitly call it out somewhere (in docs or licensing?)

Reply:
Are we going to double-count users if they view a dashboard anonymously then log in? I suspect that to prevent that case we may want to clear any anonymous entry when we see an authed user